dali/pipeline/operators/reader/reader_op.h

```diff
@@ -169,6 +171,56 @@ class DataReader : public Operator<Backend> {
   }
 
  protected:
+  void ParseIfNeeded(const Tensor<CPUBackend>& tensor, SampleWorkspace* ws) {
+    using OutputCache = std::unordered_map<std::string, std::vector<Tensor<CPUBackend>>>;
+    static OutputCache output_cache;
+    static std::mutex output_cache_mutex;
+
+    const auto& source_info = tensor.GetSourceInfo();
+    const auto should_skip_sample = tensor.ShouldSkipSample();
+    const std::size_t num_outputs = ws->NumOutput();
+
+    if (should_skip_sample) {
+      std::lock_guard<std::mutex> lock(output_cache_mutex);
+      auto it = output_cache.find(source_info);
+      DALI_ENFORCE(it != output_cache.end(),
+        "Image `" + source_info + "` should be in cache (cache size: "
+        + std::to_string(output_cache.size()) + ")");
+      auto& cached_outputs = it->second;
+      DALI_ENFORCE(cached_outputs.size() == num_outputs,
+        "Unexpected number of outputs");
+      for (std::size_t i = 0; i < cached_outputs.size(); i++) {
+        auto& output = ws->Output<CPUBackend>(i);
+        output.Copy(cached_outputs[i], 0);
+      }
+      return;
+    }
+
+    parser_->Parse(tensor, ws);
+
+    if (skip_cached_images_) {
```

Comment:
I'm not sure to get why you are parsing the tensor even when skip_cached_images_ == true

Reply:
`skip_cached_images_` tells that we want to use cache and store parsed sample in it (because it is not in cache already), while `should_skip_sample` tells if sample is in cache, load it and return.

Reply:
Alright that makes sense
Maybe it would be nice to change the name to something like `should_cache_images_`

Reply:
`skip_cached_images` is an option to the reader. It is telling the reader to skip reading images that are already in cache. Here I am just using its value to determine whether I should cache other outputs (e.g. labels) at the reader level

Reply:
Fair enough